Use debian trixie release for base node images

derekpierre opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [x] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [x] 2
- [ ] 3

**What this does:**

Based over #3647 .

Moves up changes from #3643 from `7.7.x` to `7.6.1.y`.

Switch from older debian bullseye to newer debian trixie, which reduces vulnerabilities for our docker images. Further process improvements can be made but this is a quick and relatively impactful win.

Improvements:
- rust-python image improvement:

<img width="910" height="233" alt="Screenshot 2025-09-04 at 12 35 24 PM" src="https://github.com/user-attachments/assets/1266f503-3f63-4df5-bbcb-328b157848d1" />

- corresponding nucypher image (`scout`) over the latest 7.6.0 image:

<img width="904" height="377" alt="Screenshot 2025-09-04 at 1 28 29 PM" src="https://github.com/user-attachments/assets/c34daed1-667f-4fe3-8d7b-e04103d01352" />

Updates:
- Use python 3.12.11
- `adduser` must be explicitly installed on debian trixie.

derekpierre committed:
Use debian trixie as the base rust image for our rust-python image.

derekpierre committed:
Use updated rust-python:3.12.11 that is based on debian trixie for `nucypher` docker image builds.

derekpierre committed:
Add feature newsfragment for #3643.

derekpierre committed:
Set max chunk size to 1,000 instead of 10,000.

derekpierre committed:
Handle issues during event scanning better with block range adjustments, retries and delays.
The EventScanner has to be adaptable since rpc endpoints can experience issues retrieving events due to rate/size limit/block range restrictions.

derekpierre committed:
Re-use chunk size that successfully worked after decrease.

derekpierre committed:
Add additional check and adjustment when encountering alchemy free tier event logs limits.
Add a few env variables to provide us with some additional flexibility.

Co-authored-by: Manuel Montenegro <manuel@nucypher.com>
Co-authored-by: KPrasch <kieranprasch@gmail.com>

derekpierre committed:
Add enforcement of min/max scan chunk size for EventScanner.
Updated unit tests accordingly.

derekpierre committed:
Add tests for _get_logs processing including http errors, retries and alchemy free tier checks.

manumonti committed:
Apply ruff formatting

manumonti committed:
Enable json error test

manumonti committed:
Add test for get logs when connection error

derekpierre committed:
Rename `i` variable to `attempt` for clarity.

derekpierre committed:
Add bugfix newsfragment for #3647.

derekpierre committed:
One environment variable to rule them all (for dkg ritual). Allow setting of min chunk size specifically for DKG rituals that basically allows overriding min chunk size for event scanning.

derekpierre committed:
Used named constants for default parameters used by EventScanner.
Update method documentation.

derekpierre committed:
Update event scanning tests to use higher level call to scan_chunk() instead of lower level _get_log() call; this improves test coverage.
Add test to ensure that alchemy free tier is only retried once regardless of number of retries, since retrying doesn't change the chunk size.

derekpierre committed:
Rename parameters and constants for clarity.

derekpierre committed:
Improve error messaging for chunk size decrease factor.
Add value check for chunk size increase factor.
Add testst.

derekpierre committed:
Add test ensuring that reduced chunk size is used for scanning the remaining chunks in the block range.

derekpierre committed:
Ensure that environment variables are casted to int.

derekpierre committed:
Rename feature newsfragment from #3643 to #3648.

derekpierre committed:
Update newsfragment based on David's suggestion in original PR.

Co-authored-by: David Núñez <david@nucypher.com>

manumonti reviewed (approved):
LGTM! 👌

derekpierre commented:
FYI Latest commit in this PR is running on `tapir`.

Muhammad-Altabba commented on nucypher/utilities/events.py:
```diff
@@ -69,6 +92,29 @@ def write_events_to_csv_file(csv_file: Path,
+    try:
+        # very specific error case
+        rpc_response = http_error.response.json()
+        is_http_400 = http_error.response.status_code == HTTPStatus.BAD_REQUEST
```
Was this error code verified to be the one sent back by Aclchemy for this specific scenario?

In the documentation it seems the status code for such a situation would be 403.
In their docs the HTTP Code 400 Meaning is: `Bad Request: The request is invalid (e.g incorrect format)` . And the Example Solution is: `Verify your request body and format is correct.`.
While HTTP Code 403 Meaning is: `Forbidden: the server has received the request successfully, but is intentionally refusing to process it due to some missing criteria.`. And the Example Solution is: `Upgrade tiers to increase capacity - Create a new API key - Update allowlist to permit request origin.`.

https://www.alchemy.com/docs/reference/error-reference#http-status-codes

Muhammad-Altabba replied:
I saw that 400 seems to be confirmed working as in https://github.com/nucypher/nucypher/pull/3647#issuecomment-3309683879.

However, having integration test is still helpful. If Alchemy later changed something in this regards, the integration test will fail and we will get to know about the issue faster, I think. Especially that the error message to contain the string `"Free tier"` is not mentioned in their docs: https://www.alchemy.com/docs/reference/error-reference#http-status-codes. And this may indicate a sudden change later similar to how they changed it now without an announcement.
The integration test just need to pass call Alchemy free tier url with the parameters to cause it to return the same `Free tier` error.

derekpierre replied:
✔️  - addressed in #3647 via https://github.com/nucypher/nucypher/pull/3647/commits/babcbf755c50d739cf505a1d6351b11f96ac10e0.

Muhammad-Altabba reviewed (dismissed):
Can this have integration test that at least can run locally, if not at the CD/CI?
Also added a comment as the status code does not seem to go along the docs.

derekpierre commented on nucypher/utilities/events.py:
```diff
@@ -360,9 +446,89 @@ def scan(
+            logs = web3.eth.get_logs(event_filter_params)
+            return logs, to_block_to_use
+        except HTTPError as http_error:
+            logger.warn(
+                f"eth_getLogs API call failed for range {from_block} - {to_block_to_use} ({to_block_to_use - from_block} blocks) on attempt {attempt + 1}/{max_retries}: {http_error}"
+            )
```
The logger is a `twisted` logger not a Python logger. The `warn` call in `twisted` is not deprecated.

derekpierre committed:
Add acceptance test that runs using an actual alchemy polygon rpc endpoint that is under free tier.
The test only runs if the appropriate env variable is set - it is automatically set for CI.
